### PR TITLE
Refactor solcx benchmark, add compositional field benchmark, and add active particle benchmark

### PR DESCRIPTION
**Pull request by hlokavarapu**

### CODE REVIEW

**Comment by gassmoeller** [comment]
Why did you move all of the stuff from the .cc file to the header file, and what are the 500 additional lines in the header file? Are they important? Because if you would have left everything in the .cc file the particle and compositional fields plugin could have been as simple as:
```
#include "../solcx.cc"
```
Or is there something I overlooked?

**Comment by hlokavarapu** [comment]
After talking to Timo, he convinced me that including a .cc file isn't exactly clean. Essentially, those 500 extra lines are the analytical solution moved from source to include.

File: benchmarks/solcx/compositional_fields/solcx_compositional_fields.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#ifndef ASPECT_SOLCX_COMPOSITIONAL_FIELDS_H
+#define ASPECT_SOLCX_COMPOSITIONAL_FIELDS_H
+
+#include "../benchmarks/solcx/solcx.h"
```
Comment (tjhei) [inline benchmarks/solcx/compositional_fields/solcx_compositional_fields.h]: ``#include "../solcx.h"``

File: benchmarks/solcx/solcx.cc [hunk 2]
```diff
@@ -1,2786 +1,4 @@
...
-        t657 = t30 * t304;
-        t660 = t30 * t146;
-        t665 = t25 * t17;
-        t668 = t50 * t424;
-        t671 = -0.4e1 * t321 * t90 * t146 - 0.6e1 * t643 * t232 + 0.8e1 * t182 * t209 - 0.16e2 * t648 * t510 + 0.8e1 * t651 * t136 + 0.8e1 * t89 * t4 * t100 - 0.2e1 * t374 * t657 - 0.8e1 * t648 * t660 + 0.8e1 * t130 * t328 + 0.2e1 * t665 * t99 + 0.2e1 * t346 * t668;
-        t672 = t90 * t424;
-        t676 = t120 * t536;
-        t680 = t436 * t41;
-        t688 = t366 * t67 * t440;
-        t696 = xc *
```
Comment (tjhei) [inline benchmarks/solcx/solcx.cc]: just ``#include "solcx.h"``

File: benchmarks/solcx/compositional_fields/solcx_particles.prm [hunk 3]
```diff
@@ -0,0 +1,131 @@
+# A description of the SolCX benchmark for which a known solution
+# is available. See the manual for more information.
```
Comment (tjhei) [inline benchmarks/solcx/compositional_fields/solcx_particles.prm]: can you add a sentence here (that you represent density? using particles)

**Comment by hlokavarapu** [comment]
@tjhei 

Due to the refactoring, the following code change below may have been overlooked. 

I removed the function call to AssertDimension because if we use compositional fields, then the number of components is larger than 4.

```
          virtual void vector_value (const Point< dim >   &p,
                                     Vector< double >   &values) const
          {
            AssertDimension(values.size(), 4);

            double pos[2]= {p(0),p(1)};
            double total_stress[3], strain_rate[3];
            double eta_A=1.0;
            double eta_B=eta_B_;

            // call the analytic function for the solution with a zero
            // background density
            AnalyticSolutions::_Velic_solCx
            (pos,
             eta_A, eta_B,
             0.5, 1,
             &values[0], &values[2], total_stress, strain_rate );

            // then add the background pressure to the value we just got
            values[2] += (0.5-p[1])*background_density;
          }
```

Otherwise, this PR is ready to be reviewed.

File: benchmarks/solcx/compositional_fields/solcx_particles.prm [hunk 4]
```diff
@@ -0,0 +1,133 @@
+# A description of the SolCX benchmark using active particles
```
Comment (gassmoeller) [inline benchmarks/solcx/compositional_fields/solcx_particles.prm]: Why is there a solcx_particles file in the compositional_fields folder? Did this happen by accident?

**Comment by gassmoeller** [comment]
Oh and your change above seems reasonable, but what does the 4 stand for? 2 velocity components + pressure + temperature? Then maybe it would make more sense to do a `AssertDimension(values.size(),4+n_compositional_fields);`, or something similar?

File: benchmarks/solcx/compositional_fields/solcx_particles.prm (hunk 4 above)
Comment (hlokavarapu) [inline benchmarks/solcx/compositional_fields/solcx_particles.prm]: Both solcx_particles.prm and solcx_compositional_fields.prm load the same built library libsolcx_compositional_field.so. This was intentional due to this reason.

Comment (gassmoeller) [inline benchmarks/solcx/compositional_fields/solcx_particles.prm]: But did we not create the new directory `compositional_fields` for exactly the reason to separate the compositional field case from the particle case? And the particle case in the `particle` subfolder would get its own shared library that also just needs to include the files from `benchmarks/solcx/`? At least that is how I remember this, did I mix that up in my mind?

**Comment by gassmoeller** [comment]
Not yet ready, but nearly. @hlokavarapu let us know when you are finished. Disregard my comment about the filenames, the subfolder actually describes a different setup, and the different files in the folder cover different implementations of the same benchmark (one with fields, one with particles).

**Comment by hlokavarapu** [comment]
Error table for SolCx checked out from the master branch (base case):

```
       Q2_P-1
h
3     Errors u_L1, p_L1, u_L2, p_L2: 8.912033e-06, 1.082874e-03, 1.316319e-05, 1.478426e-03
4     Errors u_L1, p_L1, u_L2, p_L2: 1.120287e-06, 2.717025e-04, 1.662154e-06, 3.697479e-04
5     Errors u_L1, p_L1, u_L2, p_L2: 1.401929e-07, 6.791438e-05, 2.082626e-07, 9.223523e-05
6     Errors u_L1, p_L1, u_L2, p_L2: 1.753057e-08, 1.697607e-05, 2.604776e-08, 2.302109e-05
7     Errors u_L1, p_L1, u_L2, p_L2: 2.191478e-09, 4.243610e-06, 3.256405e-09, 5.749812e-06
```

Error table for refactored SolCx

```
       Q2_P-1
h
3     Errors u_L1, p_L1, u_L2, p_L2: 8.912033e-06, 1.082874e-03, 1.316319e-05, 1.478426e-03
4     Errors u_L1, p_L1, u_L2, p_L2: 1.120287e-06, 2.717025e-04, 1.662154e-06, 3.697479e-04
5     Errors u_L1, p_L1, u_L2, p_L2: 1.401929e-07, 6.791438e-05, 2.082626e-07, 9.223523e-05
6     Errors u_L1, p_L1, u_L2, p_L2: 1.753057e-08, 1.697607e-05, 2.604776e-08, 2.302109e-05
7     Errors u_L1, p_L1, u_L2, p_L2: 2.191478e-09, 4.243610e-06, 3.256405e-09, 5.749812e-06
```

I am confident that I have not broken SolCx. Please do not quote me on that.

**Pull request merged**